test/Jacobian_test.cpp

```diff
@@ -0,0 +1,394 @@
+/*******************************************************************************
+
+   Tests for jacobian computation methods.
+
+   This file is part of XAD, a comprehensive C++ library for
+   automatic differentiation.
+
+   Copyright (C) 2010-2024 Xcelerit Computing Ltd.
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU Affero General Public License as published
+   by the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Affero General Public License for more details.
+
+   You should have received a copy of the GNU Affero General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+******************************************************************************/
+
+#define _USE_MATH_DEFINES
+#include <XAD/Jacobian.hpp>
+#include <XAD/XAD.hpp>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <functional>
+#include <list>
+#include <numeric>
+#include <vector>
+
+TEST(JacobianTest, SimpleAdjoint)
+{
+    typedef xad::adj<double> mode;
+    typedef mode::tape_type tape_type;
+    typedef mode::active_type AD;
+
+    tape_type tape;
+
+    std::vector<AD> x = {3.0, 1.0};
+
+    // f(x) = [ x[0] + sin(x[1]), x[1] + sin(x[0]) ]
+    auto foo = [](std::vector<AD> &x) -> std::vector<AD>
+    { return {x[0] + sin(x[1]), x[1] + sin(x[0])}; };
+
+    std::vector<std::vector<AD>> expected_jacobian = {{1.0, cos(x[1])},  //
```

Comment:
The expected values should be just `vec<vec<double>>`. This actually applies to all the Hessian and Jacobian tests. You could do that in the tests as

```c++
std::vector<double> xd = {3.0, 1.0};
std::vector<AD> x(begin(xd), end(xd));

...
```

And to confirm that output type, for the tests, better not use `auto` but specify the return type explicitly, so it would flag an error if it's not the right type.